76 Add support for asyncomplete

wellle opened the pull request:
Close #76

Edit: You don't need anything to enable it, it gets registered with default options automatically.

>To enable it add this to your vimrc:
>```vim
>let g:tmuxcomplete#asyncomplete_source_options = {}
>```

@kepbod @prabirshrestha: Do you want to have a look? You should be able to test it like this:
```vim
Plug 'wellle/tmux-complete.vim', { 'branch': '76-asyncomplete' }
```

prabirshrestha commented on autoload/asyncomplete/sources/tmuxcomplete.vim:
```diff
@@ -0,0 +1,38 @@
+function! asyncomplete#sources#tmuxcomplete#register(opts)
+    let l:opts = extend(extend({}, a:opts), {
+                \ 'name':      'tmuxcomplete',
+                \ 'whitelist': ['*'],
```
based on the order seems like whitelist can't be modified by the user.

prabirshrestha commented on autoload/asyncomplete/sources/tmuxcomplete.vim:
```diff
@@ -0,0 +1,38 @@
+    if a:event ==? 'stdout'
+        let a:opt['buffer'] .= join(a:data)
+
+    elseif a:event ==? 'exit'
```
I would check the exit code instead of assuming it always succeeds. When `a:event` is 'exit', `a:data` is the exit code.

prabirshrestha commented on autoload/asyncomplete/sources/tmuxcomplete.vim:
```diff
@@ -0,0 +1,38 @@
+function! asyncomplete#sources#tmuxcomplete#completor(opt, ctx)
+    " taken from asyncomplete-buffer
+    let l:kw = matchstr(a:ctx['typed'], '\w\+$')
+    let a:opt['kwlen'] = len(l:kw)
```
You should treat `a:opt` and `a:ctx` as immutable . You can create your own dict instead

prabirshrestha commented on autoload/asyncomplete/sources/tmuxcomplete.vim:
```diff
@@ -0,0 +1,38 @@
+    let a:opt['buffer'] = ''
+
+    let l:jobid = async#job#start(l:cmd, {
+                \ 'on_stdout': function('s:handler', [a:opt, a:ctx]),
```
since you are passing your options. you could add `a:ctx` as part of the option.

wellle commented:
@prabirshrestha: Thank you for the review! I think I addressed all your points and added some more improvements to the integration. Please have another look.

prabirshrestha commented:
looks good to me. Other users might want some docs on how to change the config values.

wellle commented:
Great! I added some documentation and made another change to actually register the source automatically (it's already guarded by your user autocommand) with default options.

Do you want to have a final look here? After that I'll merge and add the `asyncomplete` topic. 😉

prabirshrestha commented:
Looks good.
